File: api/java/com/udpwork/ssdb/Link.java
package com.udpwork.ssdb;

import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Link {
	private Socket sock;
	private MemoryStream input = new MemoryStream();

	public Link(String host, int port) throws Exception{
		this(host, port, 0);
	}

	public Link(String host, int port, int timeout_ms) throws Exception{
		sock = new Socket(host, port);
		if(timeout_ms > 0){
			sock.setSoTimeout(timeout_ms);
		}
		sock.setTcpNoDelay(true);
	}
	
	public void close(){
		try{
			sock.close();
		}catch(Exception e){
			//
		}
	}

	public Response request(String cmd, byte[]...params) throws Exception{
		ArrayList<byte[]> list = new ArrayList<byte[]>();
		for(byte[] s : params){
			list.add(s);
		}
		return this.request(cmd, list);
	}
	
	public Response request(String cmd, String...params) throws Exception{
		ArrayList<byte[]> list = new ArrayList<byte[]>();
		for(String s : params){
			list.add(s.getBytes());
		}
		return this.request(cmd, list);
	}

	public Response request(String cmd, List<byte[]> params) throws Exception{
		MemoryStream buf = new MemoryStream(4096);
		Integer len = cmd.length();
		buf.write(len.toString());
		buf.write('\n');
		buf.write(cmd);
		buf.write('\n');
		for(byte[] bs : params){
			len = bs.length;
			buf.write(len.toString());
			buf.write('\n');
			buf.write(bs);
			buf.write('\n');
		}
		buf.write('\n');
		send(buf);
		
		List<byte[]> list = recv();
		return new Response(list);
	}
	
	private void send(MemoryStream buf) throws Exception{
		//System.out.println(">> " + buf.printable());
		OutputStream os = sock.getOutputStream();
		os.write(buf.buf, buf.data, buf.size);
		os.flush();
	}
	
	private List<byte[]> recv() throws Exception{
		input.nice();
		InputStream is = sock.getInputStream();
		while(true){
			List<byte[]> ret = parse();
			if(ret != null){
				return ret;
			}
			byte[] bs = new byte[8192];
			int len = is.read(bs);
			//System.out.println("<< " + (new MemoryStream(bs, 0, len)).printable());
			input.write(bs, 0, len);
		}
	}
	
	private List<byte[]> parse(){
		ArrayList<byte[]> list = new ArrayList<byte[]>();
		byte[] buf = input.buf;
		
		int idx = 0;
		while(true){
			int pos = input.memchr('\n', idx);
			//System.out.println("pos: " + pos + " idx: " + idx);
			if(pos == -1){
				break;
			}
			if(pos == idx || (pos == idx + 1 && buf[idx] == '\r')){
				// ignore empty leading lines
				if(list.isEmpty()){
					idx += 1; // if '\r', next time will skip '\n'
					continue;
				}else{
					input.decr(idx + 1);
					return list;
				}
			}
			String str = new String(buf, input.data + idx, pos - idx);
			int len = Integer.parseInt(str);
			idx = pos + 1;
			if(idx + len >= input.size){
				break;
			}
			byte[] data = Arrays.copyOfRange(buf, input.data + idx, input.data + idx + len);
			//System.out.println("len: " + len + " data: " + data.length);
			idx += len + 1; // skip '\n'
			list.add(data);
		}
		return null;		
	}

}

File: api/dotnet/Link.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net;
using System.Net.Sockets;
using System.IO;

namespace ssdb
{
	class Link
	{
		private TcpClient sock;
		private MemoryStream recv_buf = new MemoryStream(8 * 1024);

		public Link(string host, int port) {
			sock = new TcpClient(host, port);
			sock.NoDelay = true;
			sock.Client.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.KeepAlive, true);
		}

		~Link() {
			this.close();
		}

		public void close() {
			if(sock != null){
				sock.Close();
			}
			sock = null;
		}

		public List<byte[]> request(string cmd, params string[] args) {
			List<byte[]> req = new List<byte[]>(1 + args.Length);
			req.Add(Encoding.Default.GetBytes(cmd));
			foreach(string s in args) {
				req.Add(Encoding.Default.GetBytes(s));
			}
			return this.request(req);
		}

		public List<byte[]> request(string cmd, params byte[][] args) {
			List<byte[]> req = new List<byte[]>(1 + args.Length);
			req.Add(Encoding.Default.GetBytes(cmd));
			req.AddRange(args);
			return this.request(req);
		}

		public List<byte[]> request(List<byte[]> req) {
			MemoryStream buf = new MemoryStream();
			foreach(byte[] p in req) {
				byte[] len = Encoding.Default.GetBytes(p.Length.ToString());
				buf.Write(len, 0, len.Length);
				buf.WriteByte((byte)'\n');
				buf.Write(p, 0, p.Length);
				buf.WriteByte((byte)'\n');
			}
			buf.WriteByte((byte)'\n');

			byte[] bs = buf.GetBuffer();
			sock.GetStream().Write(bs, 0, (int)buf.Length);
			//Console.Write(Encoding.Default.GetString(bs, 0, (int)buf.Length));
			return recv();
		}

		private List<byte[]> recv() {
			while(true) {
				List<byte[]> ret = parse();
				if(ret != null) {
					return ret;
				}
				byte[] bs = new byte[8192];
				int len = sock.GetStream().Read(bs, 0, bs.Length);
				//Console.WriteLine("<< " + Encoding.Default.GetString(bs));
				recv_buf.Write(bs, 0, len);
			}
		}

		private static int memchr(byte[] bs, byte b, int offset) {
			for(int i = offset; i < bs.Length; i++) {
				if(bs[i] == b) {
					return i;
				}
			}
			return -1;
		}

		private List<byte[]> parse() {
			List<byte[]> list = new List<byte[]>();
			byte[] buf = recv_buf.GetBuffer();

			int idx = 0;
			while(true) {
				int pos = memchr(buf, (byte)'\n', idx);
				//System.out.println("pos: " + pos + " idx: " + idx);
				if(pos == -1) {
					break;
				}
				if(pos == idx || (pos == idx + 1 && buf[idx] == '\r')) {
					idx += 1; // if '\r', next time will skip '\n'
					// ignore empty leading lines
					if(list.Count == 0) {
						continue;
					} else {
						int left = (int)recv_buf.Length - idx;
						recv_buf = new MemoryStream(8192);
						if(left > 0) {
							recv_buf.Write(buf, idx, left);
						}
						return list;
					}
				}
				byte[] lens = new byte[pos - idx];
				Array.Copy(buf, idx, lens, 0, lens.Length);
				int len = Int32.Parse(Encoding.Default.GetString(lens));

				idx = pos + 1;
				if(idx + len >= recv_buf.Length) {
					break;
				}
				byte[] data = new byte[len];
				Array.Copy(buf, idx, data, 0, (int)data.Length);

				//Console.WriteLine("len: " + len + " data: " + Encoding.Default.GetString(data));
				idx += len + 1; // skip '\n'
				list.Add(data);
			}
			return null;
		}
	}
}
